guides/unit-testing.md

```diff
@@ -0,0 +1,973 @@
+    it('registers the given token under the default network, assuming it has not changed yet', () => {
+      controller.addToken({ symbol: 'DAI' });
+
+      expect(controller.state.tokens).toStrictEqual({
+        '0x1': {
+          symbol: 'DAI',
+        },
+      });
+    });
+
+    it('registers the given token under the current network, even after it changes', () => {
+      onNetworkDidChangeListener!({ chainId: '0x42' });
+
+      controller.addToken({ symbol: 'DAI' });
+
+      expect(controller.state.tokens).toStrictEqual({
+        '0x42': {
+          symbol: 'DAI',
+        },
+      });
+    });
+  });
+});
+```
+
+However, this strategy ends up increasing the long-term maintenance cost of tests for a couple of reasons:
+
+- It makes debugging tests difficult later. Data that is critical to the purpose of a test — i.e., that is used in the execution or verification phase — may be hidden in a `beforeEach` hook, or it may be split across `beforeEach` and the test in question. This means that in order to trace data as it moves through the test, one must spend time piecing together the full picture.
+- It makes writing new tests difficult. Extracting the setup phase assumes that all tests will need to be set up in exactly the same way. But the very nature of code guarantees that this will not be the case. Inevitably, tests will involve scenarios that cause differences in behavior from the "happy path". Using `beforeEach`, in order to construct a new scenario, the test must extend what `beforeEach` is already doing, whether this is passing a different argument to the function being tested, mocking a function to return a different result, or something else. But this is not always as straightforward as it sounds. It is common to declare variables above the `beforeEach` which can be used later on in the file, but the utility of these variables is misleading, as by the time a test runs, these variables have already been used to set it up and thus cannot be reused or changed. Sometimes this creates enough of a problem that developers resort to copying the entirety of the setup code from `beforeEach` into the test, but this introduce another way that maintainers must now use to work with the tests.
+
+A function solves the problems that hooks solve without introducing the problems that they simultaneously introduce. With this pattern, tests that need to customize setup data can do so easily by passing options to the function, and this draws a clear line from the verification phase to the setup phase:
+
+✅
+
+``` typescript
+type WithControllerOptions = Partial<TokensControllerOptions>;
+
+type WithControllerCallback = (setup: {
+  controller: TokensController;
+  onNetworkDidChangeListener: (networkState: NetworkState) => void;
+}) => void | Promise<void>;
+
+type WithControllerArgs =
+  | [WithControllerOptions, WithControllerCallback]
+  | [WithControllerCallback];
+
+async function withController(...args: WithControllerArgs) {
```

Comment:
It's great that we're showing an example here of the `with___` pattern, which is super useful for setting up things that also need a "teardown" phase. But this seems like the more complex case. I more often find myself in need of setup without any teardown, which is a lot simpler to write and demonstrate.

Maybe we could include both examples, and explain that this is recommended when there are teardown steps?

Reply:
Yeah, definitely. I can add that.

Reply:
Extracted to a different section here: [1a8381b](https://github.com/MetaMask/contributor-docs/pull/38/commits/1a8381b901963910688228d92853b647f988e5e5)